modules/SerialCommHub/tiny_modbus_rtu.cpp

```diff
@@ -0,0 +1,327 @@
+    tty.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL | IXON | IXOFF | IXANY);
+    tty.c_lflag = 0;                   // no signaling chars, no echo,
+                                       // no canonical processing
+    tty.c_oflag = 0;                   // no remapping, no delays
+    tty.c_cc[VMIN] = 1;                // read blocks
+    tty.c_cc[VTIME] = 1;               // 0.1 seconds inter character read timeout after first byte was received
+
+    tty.c_cflag |= (CLOCAL | CREAD);   // ignore modem controls,
+                                       // enable reading
+    tty.c_cflag &= ~(PARENB | PARODD); // shut off parity
+    tty.c_cflag &= ~CSTOPB;            // 1 Stop bit
+    tty.c_cflag &= ~CRTSCTS;
+
+    if (tcsetattr(fd, TCSANOW, &tty) != 0) {
+        printf("Serial: error %d from tcsetattr\n", errno);
+        return false;
+    }
+    return true;
+}
+
+int TinyModbusRTU::read_reply(uint8_t* rxbuf, int rxbuf_len) {
+    struct timeval timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_usec = MODBUS_RX_INITIAL_TIMEOUT_MS * 1000; // 500msec intial timeout until device responds
+    fd_set set;
+    FD_ZERO(&set);
+    FD_SET(fd, &set);
+
+    int bytes_read_total = 0;
+    while (true) {
+        int rv = select(fd + 1, &set, NULL, NULL, &timeout);
+        timeout.tv_usec = MODBUS_RX_WITHIN_MESSAGE_TIMEOUT_MS *
+                          1000; // reduce timeout after first chunk, no uneccesary waiting at the end of the message
+        if (rv == -1) {         // error in select function call
+            perror("txrx: select:");
+            break;
+        } else if (rv == 0) { // no more bytes to read within timeout, so transfer is complete
+            break;
+        } else {              // received more bytes, add them to buffer
+            // do we have space in the rx buffer left?
+            if (bytes_read_total >= rxbuf_len) {
+                // no buffer space left, but more to read.
+                break;
+            }
+
+            int bytes_read = read(fd, rxbuf + bytes_read_total, rxbuf_len - bytes_read_total);
+            if (bytes_read > 0) {
+                bytes_read_total += bytes_read;
+                // EVLOG_info << "RECVD: " << hexdump(rxbuf, bytes_read_total);
+            }
+        }
+    }
+    return bytes_read_total;
+}
+
+/*
+    This function transmits a modbus request and waits for the reply.
+    Parameter request is optional and is only used for writing multiple registers.
+*/
+std::vector<uint16_t> TinyModbusRTU::txrx(uint8_t device_address, FunctionCode function,
+                                          uint16_t first_register_address, uint16_t register_quantity,
+                                          bool wait_for_reply, std::vector<uint16_t> request) {
+    {
+        // size of request
+        int req_len = (request.size() == 0 ? 0 : 2 * request.size() + 1) + MODBUS_BASE_PAYLOAD_SIZE;
+        uint8_t* req = new uint8_t[req_len];
```

Comment:
why don't you just use a vector (there is also vector.data() if you want to access read-only contiguess memory)? Problem with `new` is that you need to take care for exceptions until the corresponding delete call, otherwise memory leaks.  Or just use a `unique_ptr` which will also take care of releasing like
`auto req = std::make_unique<uint8_t[]>(req_len);`

Reply:
changed to unique_ptr